Add support for allocating PodCIDRs from multiple IPAM pools

gandro opened the pull request:
This adds support for IPAM pools. More specifically, this adds a new IPAM mode (`multi-pool`) which allows nodes to allocate PodCIDRs from different pools, depending on properties of the workload as defined by the user. An example where this will be useful is multi-homing (see https://github.com/cilium/cilium/issues/20129). 

Notably, this PR only adds the IPAM infrastructure which allows consumers inside Cilium-Agent to allocate PodCIDRs from non-default pools, it does not yet add any consumers which actually make use of this feature. This means that this PR is not going to be very useful on its own. At the moment, we require user to statically configure which IPAM pools a node should allocate from via the `multi-pool-node-pre-alloc` flag. This way of configuring a node's IPAM pools will likely be superseded by a custom network configuration CRD.

Overview of the current mechanism:

![Pooled IPAM CiliumNode CRD](https://user-images.githubusercontent.com/50564/234862662-7e1064bb-d2c2-4792-846c-8cb1c2ade2d1.png)

As with other IPAM modes, cilium-operator is responsible for global allocation decisions, i.e. allocating PodCIDRs to nodes. Each cilium-agent indicates its own demand in the CiliumNode CRD. Once it has received a PodCIDR for a given pool, it will then hand out individual IPs based on pool specified in the IP allocation request (typically received from the CNI plugin).

*Known Limitations*:
- Encryption and tunnel mode are not supported (blocked on https://github.com/cilium/cilium/issues/20170)
- In the current form, this will not set up local node routes, since the setup of routes for alternative networks will be handled by the multi-homing feature instead. For now, we requires endpoint routes being enabled.
- Cilium's built-in legacy masquerading does not support multiple pod CIDRs (c.f. https://github.com/cilium/cilium/issues/22273). To work around this, use either BPF masquerading if there is only one single native-routing-cidr. Otherwise, use ip-masq-agent (iptables or BPF work both)
- This version of the PR requires that pools are statically defined in the operator config, using the newly added `ipam.operator.multiPoolMap` Helm value. Support for managing via CRDs instead will be implemented in a follow-up PR (c.f. [`ippool-crd`](https://github.com/tklauser/cilium/tree/pr/tklauser/ippool-crd) branch here)

Joint work with @tklauser

*Review per commit*

Meta-issue: #24764

gandro committed:
clusterpool-v2beta: Refactor CIDR release and status

This commit refactors the internals of the clusterpool-v2beta
implementation a bit, to prepare the code for multi-pool, which will
share some of the code. We intend to deprecate v2beta in favor of
multi-pool in a future release, but for now there is some code sharing
involved.

The first change in this commit is adding a method to the nodeUpdater
interface. It's not used by the old implementation, but will be used in
a subsequent commit.

The other change is changing the behavior of the `status()` method,
which the new code will implement in a different manner. To accommodate
for that, the allocation and release thresholds are moved out of the
pool definition itself, and are instead passed whenever the new status
(and thus new allocations and releases) are to be computed.

The status method (now called clusterPoolV2Beta1Status) now also
performs CIDR release in a separate helper function, instead of inline
while generating the status. This simplifies the code further.

Co-authored-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
crd: Add IPAM pool fields to CiliumNode CRD

This commit adds a new `.Spec.IPAM.Pools` field to the CiliumNode CRD.
It will be used by the new IPAM pools feature, where the Cilium-Agent
running on a particular node will write its allocation requests into to
the `Requested` field, which are then fullfilled by the operator by
assigning CIDRs to the node and writing them into the `Allocated` field.
The agent is allowed to remove CIDRs from that field if it no longer
needs them. Every request and allocation is tracked per pool. The fields
will be sorted by pool, and each pool will have the CIDRs sorted as
well, with IPv4 CIDRs coming first, and IPv6 CIDRs coming second.

The actual operator and agent logic which uses those fields will follow
in subsequent commits.

Co-authored-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

tklauser committed:
helm: Add values to configure IP pools in operator

Handles IPv6 and mixed IPv4/IPv6 pools as well.

Example:

ipam:
  mode: multi-pool
  operator:
    multiPoolMap:
      mars:
        ipv4CIDRs:
        - 172.16.0.0/16
        - 172.17.0.0/16
        ipv4MaskSize: 24
      jupiter:
        ipv6CIDRs:
        - f00d::/72
        - fc00::/64
        ipv6MaskSize: 96
      saturn:
        ipv4CIDRs:
        - 192.168.0.0/24
        - 192.168.1.0/24
        ipv4MaskSize: 28
        ipv6CIDRs:
        - d00d::/70
        - dc00::/62
        ipv6MaskSize: 72

This should result in the following ConfigMap entry:

  multi-pool-map: |
    jupiter=jupiter=ipv6-cidrs:f00d::/72,fc00::/64;ipv6-mask-size:96
    mars=ipv4-cidrs:172.16.0.0/16,172.17.0.0/16;ipv4-mask-size:24
    saturn=ipv4-cidrs:192.168.0.0/24,192.168.1.0/24;ipv4-mask-size:28saturn=ipv6-cidrs:d00d::/70,dc00::/62;ipv6-mask-size:72

Co-authored-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

gandro committed:
agent: Add multi-pool IPAM allocator

This commit adds the agent side of the multi-pool IPAM allocator.  Most
of the logic is shared with the existing clusterpool-v2beta allocator.
The main differences are that multi-pool supports multiple pools, and
explicitly announces the number of needed IPs to the operator.  That
number of needed IPs is computed as the number of allocated IPs, rounded
up the next multiple of a per-pool pre-allocate value. This value will
also be used to determine if the node has excess CIDRs, in which case
the node will attempt to release any unused CIDRs back into the IPAM
pool by removing it from its CiliumNode CRD.

The pools that the agent will request from the operator are statically
configured in via the newly added hidden `multi-pool-node-pre-alloc`
flag. This flag is intended to be temporary. The IPAM pools a node
requires will instead be derived from the networks the node is attached
to. That feature to support multi-homed nodes will be added in a
subsequent PR.

Co-authored-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
operator: Add core logic for IPAM pools

This commit adds the core logic for allocating pod CIDRs from specific
pools to nodes. The logic is not yet hooked up with the CiliumNode
watcher, that will follow in a subsequent commit.

The main idea behind the `PoolAllocator` is that it tracks both the
available IP pools, but also which node has allocated which CIDR from
which pool. Each pool is implemented as a slice of `CIDRAllocators`
(based on the `github.com/cilium/ipam/cidrset` package), which means it
manages the free-lists for each pool, but does not keep track of who
allocated the CIDR. This is done separately to easily be able to obtain
a list of CIDRs allocated to a node, without having to iterate over all
pools.

As with other IPAM allocators, a main concern is state restoration after
the operator has been restarted. The approach we take here is similar to
the one used by clusterpool (v1): In an initial phase (indicated by
`PoolAllocator.ready=false`), the watcher passes in all known
CiliumNodes to `AllocateToNode`. This will cause the `PoolAllocator` to
mark any previously (before the operator restart) allocated CIDR as
occupied. Once all CiliumNodes have been observed, the flip is switched
via `RestoreFinished`, at which point the watcher will have to invoke
`AllocateToNode` again for every node, allowing new CIDRs now to be
allocated to that node if necessary. This ensures that we make sure we
know what CIDRs have been previously allocated before we hand out new
CIDRs.

Co-authored-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
operator: Hook up multi-pool IPAM allocator

This commit implements the necessary interfaces that need to be
implemented by an IPAM allocator in the operator around the
`PoolAllocator` addded in the previous commit. It adds a `NodeHandler`
which receives CiliumNode events from the K8s watcher and forwards them
to the PoolAllocator. The `NodeHandler` is also responsible to sync the
the in-memory state (of currently allocated CIDRs) to the CiliumNode CRD
in the cluster using the `CiliumNodeGetterUpdater` interface.

The list of available IPAM pools is configured by an operator flag added
by this commit. In the future, IPAM pools will be populated via a
IPPools CRD, but for now we only support this static configuration
behind a hidden flag.

Finally, we hookup up this implementation as an available IPAM allocator
in the operator.

Co-authored-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Tobias Klauser <tobias@cilium.io>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

zacharysarah reviewed (approved):
LGTM from a docs perspective.

gandro committed:
node/types: Include IPAM pools podCIDRs in alloc CIDR list

This commit appends any CIDR present in `Spec.IPAM.Pools.Allocated` to
as AllocCIDR of that node. This is needed to make the
`auto-direct-node-routes` feature work, which sets routes to the node
based on its AllocCIDRs.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

squeed commented on pkg/ipam/pool.go:
nit: the docblock is wrong now

squeed commented on pkg/ipam/pool.go:
```diff
@@ -31,12 +31,10 @@ import (
-	releaseThreshold    int
+	mutex        lock.Mutex
+	ipAllocators []*ipallocator.Range
+	released     map[string]struct{}
```
from discussion: it would be nice to know what the map keys are here.

squeed commented on pkg/ipam/clusterpool_v2beta2.go:
```diff
@@ -0,0 +1,474 @@
+			continue
+		case <-timer.After(5 * time.Second):
+			log.WithFields(logrus.Fields{
+				logfields.HelpMessage: "Check if cilium-operator pod is running and does not have any warnings or error messages.",
```
add pool name here too

squeed commented on pkg/ipam/clusterpool_v2beta2.go:
```diff
@@ -0,0 +1,474 @@
+		c.mutex.Lock()
+		switch family {
+		case IPv4:
+			if p, ok := c.pools[poolName]; ok && p.v4 != nil && p.v4.hasAvailableIPs() {
```
from meeting: what happens if the pool happened to be full on startup?

squeed replied:
answer: we haven't marked any existing reservations yet, so this is OK (but confusing)

gandro committed:
clusterpool-v2beta: Rename confusingly name variable

This commit renames a variable which has a confusing name. It tracks
both newly added and existing allocators, so let's rename it.

This commit contains no functional changes.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipam: Document and test neededIPCeil

This commit documentes the behavior of neededIPCeil and adds a unit test
to assert it too. This commit also adds a division-by-zero check, but
otherwise should not contain any user visible functional changes.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

tklauser commented on operator/cmd/flags.go:
```diff
@@ -39,6 +39,11 @@ func init() {
 	option.BindEnv(Vp, operatorOption.IPAMInstanceTags)
 
+	flags.Var(option.NewNamedMapOptions(operatorOption.IPAMMultiPoolMap, &operatorOption.Config.IPAMMultiPoolMap, nil), operatorOption.IPAMMultiPoolMap,
+		"IP pool definitions in the form <pool>=ipv4-cidrs:<cidr>,[<cidr>...];ipv4-mask-size:<size> (multiple pools van also be passed by repeating the CLI flag)")
```
Nit:

```suggestion
		"IP pool definitions in the form <pool>=ipv4-cidrs:<cidr>,[<cidr>...];ipv4-mask-size:<size> (multiple pools can also be passed by repeating the CLI flag)")
```

tklauser commented on pkg/ipam/allocator/multipool/pool_allocator.go:
```diff
@@ -0,0 +1,402 @@
+		return fmt.Errorf("cannot allocate from non-existing pool: %s", sourcePool)
+	}
+
+	log.WithFields(logrus.Fields{
+		"targetNode": targetNode,
+		"sourcePool": sourcePool,
+		"family":     family,
+		"toAllocate": toAllocate,
+	}).Debug("allocating cidr")
```
This log line might be confusing because it will also log for negative `toAllocate`:

```
level=debug msg="allocating cidr" family=ipv4 sourcePool=default subsys=ipam-allocator-clusterpool-v2 targetNode=kind-worker toAllocate=-26
```

but the loop below will be skipped and nothing allocated. Probably better to skip the log message if `toAllocate <= 0`.

gandro replied:
Fixed by adding a `toAllocate <= 0` check in allocate

tklauser reviewed (approved):
LGTM, two small remarks inline.

Also verified using the command given in https://github.com/cilium/cilium/pull/22762#issuecomment-1543938640 that the pool get created properly:

```
% ks logs cilium-operator-5dd5f6dd8c-5bkvl
[...]
level=info msg="  --multi-pool-map='default=ipv4-cidrs:10.0.0.0/8;ipv4-mask-size:27'" subsys=cilium-operator-generic
[...]
level=debug msg="found pool definition" pool="ipv4-cidrs:10.0.0.0/8;ipv4-mask-size:27" pool-name=default subsys=ipam-allocator-clusterpool-v2
level=debug msg="adding pool" ipv4-cidrs="[10.0.0.0/8]" ipv4-masksize=27 ipv6-cidrs="[]" ipv6-masksize=0 pool-name=default subsys=ipam-allocator-clusterpool-v2
[...]
level=debug msg="allocating cidr" family=ipv4 sourcePool=default subsys=ipam-allocator-clusterpool-v2 targetNode=kind-worker toAllocate=64
```

christarazi commented on pkg/ipam/types/types.go:
```diff
@@ -51,6 +51,51 @@ type AllocationIP struct {
+	// CIDRs contains a list of pod CIDRs currently allocated from this pool
+	//
+	// +optional
+	CIDRs []string `json:"cidrs,omitempty"`
```
We can validate these CIDRs using

```
+kubebuilder:validation:Format=cidr
```

gandro replied:
Fixed by introducing a newtype with that validation annotation.

christarazi commented on pkg/ipam/types/types.go:
```diff
@@ -159,6 +209,22 @@ type IPAMStatus struct {
+	// pool
+	//
+	// +optional
+	IPv4Addrs int `json:"ipv4-addrs,omitempty"`
```
Nit: in CRD YAMLs, we have a preference in the codebase for camelCase

squeed replied:
yeah, camelCase is the convention for k8s yaml / json documents.

gandro replied:
Almost all other CiliumNode fields are also `kebab-case` unfortunately, see for example the ENISpec used to manage the ENI IPAM mode: https://github.com/cilium/cilium/blob/c5cbf403dbe355ecbb80dfc8d7a8ed4da45c43bd/pkg/aws/eni/types/types.go#L18

I made `ipv4-addrs` and `ipv6-addrs` kebab-case to be consistent with the existing IPAM fields.

If you feel strongly that we should make newer fields camelCase, I can change it, but otherwise I'll leave it kebab-case.

christarazi replied:
Oh that's unfortunate. In that case, better to be consistent with kebab-case.

christarazi commented on pkg/ipam/types/types.go:
```diff
@@ -51,6 +51,51 @@ type AllocationIP struct {
+
+	// Needed indicates how many IPs out of the above Pool this node requests
+	// from the operator. The operator runs a reconciliation loop to ensure each
+	// node always has enough PodCIDRs allocated in each pool to fulfil the
```
Typo: `fulfill`

christarazi commented on go.mod:
```diff
@@ -92,6 +92,7 @@ require (
 	go.uber.org/goleak v1.2.1
 	go.uber.org/multierr v1.11.0
 	go.universe.tf/metallb v0.11.0
+	go4.org/netipx v0.0.0-20230303233057-f1b76eb4bb35
```
TIL: This seems like a nice package that we can use throughout and remove the duplicated functions in `pkg/ip`. Makes more sense as a followup.

tklauser replied:
Yep, the intention is to replace most of the `netip` convenience funcs in `pkg/ip` by the ones in this package in a follow-up PR (cf. #24246).

christarazi commented on pkg/ipam/allocator/multipool/pool.go:
```diff
@@ -0,0 +1,112 @@
+			return netip.Prefix{}, err
+		}
+
+		prefix, ok := netipx.FromStdIPNet(ipnet)
```
JFYI and related to my previous comment about this package: this is the only remaining function AFAICT that's missing from our helpers in `pkg/ip`.

gandro replied:
Good catch! I'll leave that for a follow-up PR and left a comment here: https://github.com/cilium/cilium/issues/24246#issuecomment-1547961878

christarazi commented on pkg/ipam/allocator/multipool/pool.go:
```diff
@@ -0,0 +1,112 @@
+			return err
+		}
+		if !allocated {
+			return nil // not an error to release a cidr twice
```
Can this actually occur or is this more of defensive coding?

If it can occur, I'm curious to understand what might trigger more than one release.

gandro replied:
This is defensive coding. I cannot come up with a scenario where we would release an IP twice.

christarazi commented on pkg/ipam/multipool.go:
```diff
@@ -0,0 +1,480 @@
+	ctx := context.Background()
+	for pool := range m.preallocMap {
+		if m.conf.IPv4Enabled() {
+			m.waitForPool(ctx, IPv4, pool)
```
This context isn't effective. Is this intended?

gandro replied:
Ah, good catch! The current _behavior_ is intended, i.e. IPAM will block the agent forever if it is not ready. This is consistent with other IPAM modes.

But I think it still makes sense to have a per-pool context, such that we get an error message for each pool which is not ready. I'll fix that.

christarazi reviewed (commented):
Mostly LGTM, nice work! I left a few minor comments.

I have some questions as well:
1. Should we just announce (and document) deprecation of clusterpool-v2beta?
2. Should we document the known limitations of multi-pool as described in the PR description?
3. I don't see any end-to-end tests and I presume that you & Tobias have manually validated the feature locally. I'd like to see coverage of it sooner rather than later so that we can iron out any bugs before it's released. I wonder if we should configure the kind workflow to run with this new IPAM mode. What are your thoughts on the e2e testing?

squeed reviewed (approved):
Looks great!

rolinh reviewed (approved):
Vendor changes lgtm.

christarazi reviewed (approved):
It would be good to have an issue to track the followups discussed in https://github.com/cilium/cilium/pull/22762#issuecomment-1548144692
